internal/controllers/synthesis/slicecleanup.go

```diff
@@ -98,7 +104,7 @@ func shouldReleaseFinalizer(comp *apiv1.Composition, slice *apiv1.ResourceSlice)
 	if comp.Status.CurrentSynthesis != nil && slice.Spec.CompositionGeneration > comp.Status.CurrentSynthesis.ObservedCompositionGeneration && owner.UID == comp.UID {
 		return false // stale informer
 	}
-	return !resourcesRemain(slice) || (!synthesisReferencesSlice(comp.Status.CurrentSynthesis, slice) && !synthesisReferencesSlice(comp.Status.PreviousSynthesis, slice))
+	return comp.DeletionTimestamp != nil && (!resourcesRemain(slice) || (!synthesisReferencesSlice(comp.Status.CurrentSynthesis, slice) && !synthesisReferencesSlice(comp.Status.PreviousSynthesis, slice)))
```

Comment:
Checking the composition deletion status ensures that we don't prematurely release the resource slice if it is deleted before the composition (or before we receive the composition deletion watch event).

This can happen if apiserver's GC beats our controller to the deletion which happens although very rarely.